tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -792,17 +785,6 @@ Status StorageManagerCanonical::query_submit_async(Query* query) {
   return async_push_query(query);
 }
 
-Status StorageManagerCanonical::set_tag(
-    const std::string& key, const std::string& value) {
-  tags_[key] = value;
```

Comment:
I don't think it's a good idea to completely `tags_` as a store instead of moving it into `Context` or `ContextResources`. I guess you have considered it duplicated information, since today we add all those tags to `rest_client`'s `extra_headers` but I am not convinced that those 2 k-v stores are meant to be identical. The way I understand the current design is that `tags` are actually a subset of `extra_headers`, so that change is also a semantic change. 

As a side note, on the usefulness of `tags_` and why it would be better to keep them, I am currently working w/ @NullHypothesis from Cloud team on extending Jaeger tracing to Core C-APIs and we are using those very `tags_` to store the required tracing context (hence the attempt to introduce locking in [my other PR](https://github.com/TileDB-Inc/TileDB/pull/5022)). 
See [this](https://github.com/TileDB-Inc/TileDB-Cloud-REST/pull/4498/files#diff-40706246b989fda0dae71da3d4d596df2ef1044f30c1282da467c1902f3c4d50R159) where we get the context, and [this](https://github.com/TileDB-Inc/TileDB-Cloud-REST/pull/4498/files#diff-549557f09a22e4644d0a928521297d35c1dd2c0dd4aad63098c0ee5ef5abfbcbR30) as an example of where we set it.

Reply:
>meant to be identical

They are actually identical. I went through all the ways that `extra_headers` were modified, and every one of them went through `StorageManager`.

As for access to this data, `StorageManager::tags()` has been replaced with `RestClient::extra_headers()`. The functions return the same value. Likewise, `RestClient::add_header()` is a public member function.

Reply:
Well, there's one:
- https://github.com/TileDB-Inc/TileDB/blob/bc4ec51a689337a16302d94f6ec40ab8809704dd/tiledb/sm/rest/rest_client.cc#L344

Also if you rebase this PR , there's also this check in the current code before mirroring to headers: 
```
  if (resources_.rest_client() != nullptr)
    throw_if_not_ok(resources_.rest_client()->set_header(key, value));
```

Are there cases where we don't have a `rest_client` configured? If yes then we'd still want the tag API to work through `tags_`.

Reply:
Also if we just keep `extra_headers_` , could we add a mutex to guard modifications?

Reply:
> Are there cases where we don't have a `rest_client` configured

This PR hasn't merged yet because of fundamental build problems. I'm rewriting the class.
* The context will always have a `rest_client` instance.
* The member variable `extra_headers_` is always defined regardless of the whether the serialization is enabled or not.

> Also if we just keep `extra_headers_` , could we add a mutex to guard modifications?

Could the extra headers you want be specified at construction time? There's a different PR up right now allowing loading them at construction from the configuration argument.